kyo-data/shared/src/test/scala/kyo/RecordTest.scala

```diff
@@ -0,0 +1,473 @@
+        }
+    }
+
+    "compact" - {
+        "preserves defined fields" in {
+            val record: Record["name" ~ String & "age" ~ Int] =
+                "name" ~ "Frank" & "age" ~ 40
+            val compacted = record.compact
+
+            assert(compacted.size == 2)
+            assert(compacted.name == "Frank")
+            assert(compacted.age == 40)
+        }
+
+        "removes extra fields" in {
+            val record: Record["name" ~ String] =
+                "name" ~ "Frank" & "age" ~ 40
+            val compacted = record.compact
+            assert(compacted.size == 1)
+            assertDoesNotCompile("""
+                compacted.age
+            """)
+        }
+    }
+
+    "AsFields behavior" - {
+        "complex intersection types" in {
+            val fields = Record.AsFields["a" ~ Int & "b" ~ String & "c" ~ Int]
+            assert(fields.size == 3)
+            assert(fields.map(_.name).toSet == Set("a", "b", "c"))
+        }
+
+        "nested intersection types" in {
+            type Nested = "outer" ~ Int & ("inner1" ~ String & "inner2" ~ Boolean)
+            val fields = Record.AsFields[Nested]
+            assert(fields.size == 3)
+            assert(fields.map(_.name).toSet == Set("outer", "inner1", "inner2"))
+        }
+
+        "intersections with 4 fields" in {
+            type Fields4 = "a" ~ Int & "b" ~ String & "c" ~ Boolean & "d" ~ List[Int]
+            val fields = Record.AsFields[Fields4]
+            assert(fields.size == 4)
+            assert(fields.map(_.name).toSet == Set("a", "b", "c", "d"))
+        }
+
+        "intersections up to 22 fields" in {
+            type Fields22 =
+                "1" ~ Int & "2" ~ Int & "3" ~ Int & "4" ~ Int & "5" ~ Int & "6" ~ Int & "7" ~ Int & "8" ~ Int & "9" ~ Int & "10" ~ Int &
+                    "11" ~ Int & "12" ~ Int & "13" ~ Int & "14" ~ Int & "15" ~ Int & "16" ~ Int & "17" ~ Int & "18" ~ Int & "19" ~ Int &
+                    "20" ~ Int & "21" ~ Int & "22" ~ Int
+            assertCompiles("Record.AsFields[Fields22]")
+        }
+
+        "intersections with more than 22 fields" in pendingUntilFixed {
+            type Fields23 =
+                "1" ~ Int & "2" ~ Int & "3" ~ Int & "4" ~ Int & "5" ~ Int & "6" ~ Int & "7" ~ Int & "8" ~ Int & "9" ~ Int & "10" ~ Int &
+                    "11" ~ Int & "12" ~ Int & "13" ~ Int & "14" ~ Int & "15" ~ Int & "16" ~ Int & "17" ~ Int & "18" ~ Int & "19" ~ Int &
+                    "20" ~ Int & "21" ~ Int & "22" ~ Int & "23" ~ Int
+            assertCompiles("Record.AsFields[Fields23]")
+        }
+
+        "intersections with duplicate field names but different types" in {
+            type DupeFields = "value" ~ Int & "value" ~ String
+            val fields = Record.AsFields[DupeFields]
+            assert(fields.size == 2)
+            assert(fields.map(_.name).toSet == Set("value"))
+            assert(fields.map(_.tag).toSet.size == 2)
+        }
```

Comment:
What's the behavior for an invalid field intersection? E.g.,

`Record.AsFields[Int & String & "value" ~ Int]`

Reply:
(I assume it would be considered valid, and would be treated as a single field equivalent to `"value" ~ Int`)

Reply:
I've added a new [group of tests](https://github.com/getkyo/kyo/pull/959/commits/584d890c25f48a47c2fed7af37599c7a8400e7d6) for malformed record types. The gist:

1. It's not possible produce instances of them using the Record API. It's also not possible via type widening.
2. But users can declare methods with a malformed type since `Fields` doesn't have a type bound (maybe we could add `~` but I think we'd need variance)
3. It's possible use a record with a malformed type via unsafe type casts but not all operations work. Field selection, `toMap`, `size`, `fields`, `size`, and `&` work properly but `compact` and any other API using `AsFields` don't
4. It's not possible to materialize `AsFields` for a malformed record type

Reply:
I took a shot at adding a type bound to `Fields` but I couldn't make `AsRecord.FieldsOf` work with it